File: frontend/passes/reachability.go
package passes

import "github.com/mattcarp12/maml/frontend/mir"

func EliminateDeadFunctions(prog *mir.Program) {
	if prog == nil {
		return
	}

	liveSet := make(map[string]bool)
	var worklist []string

	// 1. Mark entry points
	entryPoints := []string{"main",
		"maml_coro_runtime_init",
		"maml_alloc",
		"maml_free",
	}
	for _, ep := range entryPoints {
		worklist = append(worklist, ep)
		liveSet[ep] = true
	}

	// Helper to find a function by name
	getFunc := func(name string) *mir.Function {
		for i := range prog.Functions {
			if prog.Functions[i].Name == name {
				return &prog.Functions[i]
			}
		}
		return nil
	}

	// 2. Trace the call graph
	for len(worklist) > 0 {
		funcName := worklist[0]
		worklist = worklist[1:]

		fn := getFunc(funcName)
		if fn == nil || fn.Graph == nil {
			continue // Might be an extern function with no body
		}

		// Look for CallInsts in all blocks
		for _, block := range fn.Graph.Blocks {
			for _, inst := range block.Statements {
				if call, ok := inst.(*mir.CallInst); ok {
					// The function being called is stored in the Function Value
					if reg, isReg := call.Function.(*mir.Register); isReg {
						targetName := reg.Name
						if !liveSet[targetName] {
							liveSet[targetName] = true
							worklist = append(worklist, targetName)
						}
					}
				}
			}
		}
	}

	// 3. Prune the dead functions
	var liveFunctions []mir.Function
	for _, fn := range prog.Functions {
		if liveSet[fn.Name] {
			liveFunctions = append(liveFunctions, fn)
		}
	}
	prog.Functions = liveFunctions
}

File: passes/src/reachability.cpp
#include "reachability.h"
#include "cfg.h"
#include "mir.h"
#include "sym.h"
#include <string>
#include <unordered_set>
#include <utility>
#include <variant>
#include <vector>

namespace maml::passes {

void eliminateDeadFunctions(mir::Program* prog, SymbolTable& sym)
{
    if (!prog)
        return;

    std::unordered_set<SymID> liveSet;
    std::vector<SymID> worklist;

    auto addEntryPoint = [&](const std::string& name) {
        SymID id = sym.intern(name);
        worklist.push_back(id);
        liveSet.insert(id);
    };

    // 1. Mark entry points
    addEntryPoint("main");
    addEntryPoint("maml_coro_runtime_init");
    addEntryPoint("maml_alloc");
    addEntryPoint("maml_free");

    auto getFunc = [&](SymID name) -> mir::Function* {
        for (auto& fn : prog->functions) {
            if (fn.name == name)
                return &fn;
        }
        return nullptr;
    };

    // 2. Trace the call graph
    while (!worklist.empty()) {
        SymID current = worklist.back();
        worklist.pop_back();

        mir::Function* fn = getFunc(current);
        if (!fn || !fn->graph)
            continue; // Might be an extern function with no body[cite: 3]

        // Look for CallInsts in all blocks[cite: 3]
        for (const auto& [_, block] : fn->graph->blocks) {
            for (const auto& inst : block->statements) {
                if (auto* call = std::get_if<mir::CallInst>(&inst)) {
                    // The function being called is stored in the Function Value[cite: 3]
                    if (auto* reg = std::get_if<mir::Register>(&call->function)) {
                        SymID targetName = reg->name;
                        if (liveSet.find(targetName) == liveSet.end()) {
                            liveSet.insert(targetName);
                            worklist.push_back(targetName);
                        }
                    }
                }
            }
        }
    }

    // 3. Prune the dead functions
    std::vector<mir::Function> liveFunctions;
    for (auto& fn : prog->functions) {
        if (liveSet.find(fn.name) != liveSet.end()) {
            liveFunctions.push_back(std::move(fn));
        }
    }
    prog->functions = std::move(liveFunctions);
}

} // namespace maml::passes